### ISSUE TITLE
Incompatibility with forthcoming RF 2.9

### ISSUE THREAD

**Issue by pekkaklarck** [ncbi/robotframework-pageobjects#54]
I noticed earlier that Se2Lib wasn't compatible with RF 2.9 and reported it as rtomac/robotframework-selenium2library#429. That problem was fixed, but now @rama-bornfree has commented that there are other problems. Based on the traceback and a quick look at the code, it seems that the issue is actually in this library.

I hope you have a change to analyze what's causing you problems. We have rafactored lot of code during 2.9 development, and, for example, variables have been pretty much rewritten. Please let us know ASAP if we've done something that is causing you lot of problems. We try to create RF 2.9 rc 1 this week, and have still a possibility to fix backwards incompatibilities.

We'd also like to hear if there are APIs that you'd like to be declared stable, APIs that could be improved, etc.

**Comment by zephraph** [comment]
_(Copy pasta'd from the issue I was making before @pekkaklarck ninja'd me :smile:)_

@rama-bornfree [originally reported](https://github.com/rtomac/robotframework-selenium2library/issues/429#issuecomment-119002837) in S2L that pageobjects was failing with RF 2.9b2. 

RF 2.9a3 refactored a lot of the ```robot.variables``` module and I assume ```init_global_variables``` was removed as a part of that change. The stack trace @rama-bornfree gave us referenced that. I've found it [here](https://github.com/ncbi/robotframework-pageobjects/blob/d6128aecd37cc8b311c32bf8dab84634ce6c8dc0/robotpageobjects/context.py#L7) and [here](https://github.com/ncbi/robotframework-pageobjects/blob/d6128aecd37cc8b311c32bf8dab84634ce6c8dc0/robotpageobjects/context.py#L73). 

Really all I think you need to do is replace those references with something like a ```BuiltIn``` call like I did for [S2L#431](https://github.com/rtomac/robotframework-selenium2library/pull/431). 

Hope this helps!

**Comment by hellmanj** [comment]
Thanks @pekkaklarck and @zephraph, we certainly do want to be 2.9 compatible!  @frishberg, I tried a quick fix for this in https://github.com/ncbi/robotframework-pageobjects/commit/a2de8c98679b1a672ad430b2ec35b25c2ce69b25 but there's one failed test in CI 326 which looks to be an out-of-robot issue.  Any ideas?

**Comment by frishberg** [comment]
I'll take a look...

**Comment by frishberg** [comment]
Hmm. For me, `robot.variables.scopes` doesn't exist, so I can't run any tests in your branch.

**Comment by zephraph** [comment]
yeah, `robot.variables` is a private api and they made some big changes in it to fix how variable scoping works.

**Comment by pekkaklarck** [comment]
Yes, variable related code was pretty much re-written to support dict variables, fix variable scopes, avoid global state, etc. Hopefully changes didn't break too much external code too badly. We can consider changes to make external usage easier still before 2.9 is released.

**Comment by hellmanj** [comment]
@frishberg, you have to install the development version of `robotframework` to test the change:

`pip install --upgrade --pre robotframework`

**Comment by frishberg** [comment]
Thanks. Now I see. Yeah, this is tricky. To decide where to output
screenshots, Selenium2Library's `capture_page_screenshot` calls
`_get_log_dir`:
https://github.com/rtomac/robotframework-selenium2library/blob/v1.7.0/src/Selenium2Library/keywords/_logging.py#L15
`_get_log_dir` used to check this `GLOBAL_VARIABLES` list, which we were
populating in
https://github.com/ncbi/robotframework-pageobjects/blob/master/robotpageobjects/context.py#L73
by
calling
https://github.com/robotframework/robotframework/blob/2.8.7/src/robot/variables/__init__.py#L36.
It appears that now, to retrieve the path to which to log screenshots,
Selenium2Library calls `BuiltIn().get_variables()` (
https://github.com/rtomac/robotframework-selenium2library/blob/master/src/Selenium2Library/keywords/_logging.py#L15),
which requires that an execution context exist.

In short, we were relying on code in Selenium2Library that was accidentally
not relying on executing in Robot. Now that that code uses the BuiltIn()
library, we can't assume that it will work outside Robot anymore. Perhaps
we could modify or monkey-patch `BuiltIn` so that it can retrieve variables
outside Robot. What do you think, Jared?

**Comment by hellmanj** [comment]
Do we really need to reference robot variables at all (except for S2L screenshot paths) when outside robot?  I tried removing this code altogether in https://github.com/ncbi/robotframework-pageobjects/commit/b91121dfd06bc4f036c28516c736243594becc95 and got the same test results.  So could we just patch either [`_get_log_dir`](https://github.com/rtomac/robotframework-selenium2library/blob/master/src/Selenium2Library/keywords/_logging.py#L14) or [`_get_screenshot_paths`](https://github.com/rtomac/robotframework-selenium2library/blob/master/src/Selenium2Library/keywords/_screenshot.py#L52) ?

**Comment by zephraph** [comment]
If there's anything I can do on S2L to make this easier, just let me know.

**Comment by frishberg** [comment]
It would be better if we didn't refer to robot variables, but it's the S2L code that's making the reference. Yes, we could override `_get_log_dir` or _get_screenshot_paths`. The code you removed in https://github.com/ncbi/robotframework-pageobjects/commit/b91121dfd06bc4f036c28516c736243594becc95 was actually there in the first place to make code that refers to Robot's global variables think that it was running in Robot, so it's a workaround. Perhaps there is some way of improving that, but I don't see how other than overriding the aforementioned S2L method(s).

I am happy to get started on that change. It might make things easier for us to make the variable references in S2L more generic (i.e. not assume Robot), if @zephraph and others are open to the kind of decoupling we are trying to do.

**Comment by hellmanj** [comment]
I doubt decoupling was ever meant to be in scope for S2L so I wouldn't want to impose that job there.  I think it would be simplest to override `_get_log_dir` to check the context, and when out of robot, just use the cwd (as we currently do with `po_log.txt`).

**Comment by pekkaklarck** [comment]
Overriding `_get_log_dir` ought to work, but because that's a private method it may change in the future. To prevent such problems, it would thus be best to have a designed way to configure this behavior in S2L. 

@zephraph, could S2L be changed so that if getting the log directory from Robot fails due to `RobotNotRunningError` exception, it would return a predefined default log directory? This default could be passed to `__init__` or be a public (and documented) class attribute that extending classes can set. Its default value could be just `.`.

Notice that S2L could also use Robot's `BuiltIn.get_variable_value` to resolve a single variable instead of getting all variables using `BuiltIn.get_variables`. Both of these currently raise `RobotNotRunningError` if Robot isn't running, but I could easily change the former to return the given default value instead. I'm not sure would that help too much, though, because caller can already now easily catch that exception and handle the situation as needed

**Comment by zephraph** [comment]
I can add an argument to S2L like ```screenshot_root_directory``` which will default to ```None```. If the ```RobotNotRunningError``` occurs then it'll just use the current directly. 

Sound acceptable?

**Comment by hellmanj** [comment]
Works for me, thanks!

**Comment by frishberg** [comment]
Yes, that sounds great!

**Comment by zephraph** [comment]
It's almost done I've just got to add a spec. This is a little harder to test than I thought it would be.

**Comment by zephraph** [comment]
@hellmanj @frishberg, these changes are live in 1.7.2. Let me know if you have any issues.

**Comment by hellmanj** [comment]
Looks like it works for us, thanks again!

**Comment by zephraph** [comment]
Glad to hear it. If you need any more support from the S2L side, let me know!

**Comment by zephraph** [comment]
@hellmanj, I just noticed tonight that I missed a small bug with the custom screenshot directory api. Linking to images in the log files was broken under some conditions. 

I've got a fix in [Selenium2Library/#460](https://github.com/rtomac/robotframework-selenium2library/pull/460), but I just wanted to make you aware. 

I'll probably push a hotfix either tomorrow or sometime next week. I want to wait to give someone a chance to poke it first to make sure it works in different environments. I'll keep you updated though.
